src/ast_editor.ml

```diff
@@ -0,0 +1,617 @@
+open Import
+
+exception User_error of string
+
+module Handlers = struct
+  let unpwrap = function
+    | `Ok () -> ()
+    | `Error err_msg -> show_message `Error "%s" err_msg
+
+  let w1 f x =
+    try `Ok (f x) with
+    | User_error e -> `Error e
+
+  let ws f x y =
+    match f x with
+    | `Ok f' -> (
+      try `Ok (f' y) with
+      | User_error e -> `Error e)
+    | `Error e -> `Error e
+
+  let w2 f = ws (w1 f)
+
+  let w3 f x = ws (w2 f x)
+
+  let w4 f x y = ws (w3 f x y)
+
+  let w5 f x y z = ws (w4 f x y z)
+
+  let _w6 f x y z w = ws (w5 f x y z w)
+end
+
+let read_html_file () =
+  let filename = Node.__dirname () ^ "/../astexplorer/dist/index.html" in
+  Fs.readFile filename
+
+let document_eq a b = Uri.equal (TextDocument.uri a) (TextDocument.uri b)
+
+let send_msg t value ~(webview : WebView.t) =
+  let msg = Ojs.empty_obj () in
+  Ojs.set_prop_ascii msg "type" (Ojs.string_to_js t);
+  Ojs.set_prop_ascii msg "value" value;
+  let (_ : bool Promise.t) = WebView.postMessage webview msg in
+  ()
+
+module Pp_path : sig
+  type kind =
+    | Structure
+    | Signature
+    | Unknown
+
+  val get_kind : document:TextDocument.t -> kind
+
+  val get_pp_path : document:TextDocument.t -> string
+end = struct
+  type kind =
+    | Structure
+    | Signature
+    | Unknown
+
+  let relative_document_path ~document =
+    Workspace.asRelativePath ~pathOrUri:(`Uri (TextDocument.uri document)) ()
+
+  let project_root_path () = Workspace.rootPath ()
```

Comment:
would it be possible to use dune to get the root of the project instead of vscode's workspace? The current code doesn't work in the context of a monorepo

Reply:
Hmm, but why not? Doesn't it make sense to open the monorepo such that its root matches the workspace root?

Reply:
here I'm thinking about a monorepo in a larger sense than just a dune monorepo. A repo containing multiple projects written in multiple languages with different build systems. It's convenient to have the whole thing open (search in the whole repo, check that a change in one project doesn't break the other ones, ...)

This is not a total deal breaker. It would just make the experience better. I've created https://github.com/ocamllabs/vscode-ocaml-platform/issues/873 to keep track of the idea.

Reply:
I see. Well I'm definitely not against the idea.